components/org.wso2.carbon.identity.recovery/src/main/java/org/wso2/carbon/identity/recovery/listener/SelfRegistrationCompletionListener.java

```diff
@@ -0,0 +1,317 @@
+import static org.wso2.carbon.identity.recovery.IdentityRecoveryConstants.ConnectorConfig.SIGN_UP_NOTIFICATION_INTERNALLY_MANAGE;
+import static org.wso2.carbon.identity.recovery.RecoveryScenarios.SELF_SIGN_UP;
+import static org.wso2.carbon.identity.recovery.RecoverySteps.CONFIRM_SIGN_UP;
+import static org.wso2.carbon.identity.recovery.util.SelfRegistrationUtils.getNotificationChannel;
+import static org.wso2.carbon.identity.recovery.util.SelfRegistrationUtils.getUserStoreManager;
+import static org.wso2.carbon.identity.recovery.util.SelfRegistrationUtils.handledNotificationChannelManagerException;
+import static org.wso2.carbon.identity.recovery.util.SelfRegistrationUtils.lockUserAccount;
+import static org.wso2.carbon.identity.recovery.util.SelfRegistrationUtils.maskIfRequired;
+import static org.wso2.carbon.identity.recovery.util.SelfRegistrationUtils.resolveEventName;
+import static org.wso2.carbon.identity.recovery.util.SelfRegistrationUtils.triggerAccountCreationNotification;
+import static org.wso2.carbon.identity.recovery.util.SelfRegistrationUtils.triggerNotification;
+import static org.wso2.carbon.identity.user.registration.engine.Constants.SELF_REGISTRATION_DEFAULT_USERSTORE_CONFIG;
+import static org.wso2.carbon.user.core.UserCoreConstants.APPLICATION_DOMAIN;
+import static org.wso2.carbon.user.core.UserCoreConstants.INTERNAL_DOMAIN;
+import static org.wso2.carbon.user.core.UserCoreConstants.WORKFLOW_DOMAIN;
+import static org.wso2.carbon.utils.DiagnosticLog.ResultStatus.FAILED;
+import static org.wso2.carbon.utils.DiagnosticLog.ResultStatus.SUCCESS;
+
+/**
+ * Listener to honour self registration connector configs after the successful completion of the self registration
+ * flow.
+ */
+public class SelfRegistrationCompletionListener extends AbstractFlowExecutionListener {
+
+    private static final Log LOG = LogFactory.getLog(SelfRegistrationCompletionListener.class);
+    private static final String REGISTRATION_COMPLETION_LISTENER = "registration-completion-listener";
+    private static final String TRIGGER_NOTIFICATION = "trigger-notification";
+
+    @Override
+    public int getDefaultOrderId() {
+
+        return 3;
+    }
+
+    @Override
+    public int getExecutionOrderId() {
+
+        return 3;
+    }
+
+    @Override
+    public boolean isEnabled() {
+
+        return true;
+    }
+
+    @Override
+    public boolean doPostContinue(RegistrationStep step, RegistrationContext registrationContext)
+            throws RegistrationEngineException {
+
+        if (Constants.COMPLETE.equals(step.getFlowStatus())) {
+
+            RegisteringUser user = registrationContext.getRegisteringUser();
+            String tenantDomain = registrationContext.getTenantDomain();
+            String userStoreDomain = resolveUserStoreDomain(user.getUsername());
+            Map<String, Object> loggerInputs = new HashMap<>();
+
+            try {
```

Comment:
Shall we refactor two functionalities into private methods? i.e handleAccountLock, triggerConfirmation